### PR TITLE
London | 26-ITP-January | Mouawia Elkhalifa | Sprint 2 | Data Groups

### PR DESCRIPTION
**Pull request by MouawiaElkhalifa**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the style guide

## Changelist

Implemented all Sprint 2 exercises including:
- Debug tasks (address, author, recipe)
- Implement tasks (contains, lookup, querystring, tally)
- Interpret task (invert)

All functions were tested and verified to meet the expected behavior. Edge cases such as invalid input and duplicate values were handled where required.

### CODE REVIEW

File: Sprint-2/implement/contains.js [hunk 1]
```diff
@@ -1,3 +1,9 @@
-function contains() {}
+function contains(obj, prop) {
+  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
+    return false;
+  }
+
+  return prop in obj;
```
Comment (cjyuan) [inline Sprint-2/implement/contains.js]: Consider the following two approaches for determining if an object contains a property:
```
  let obj = {}, propertyName = "toString";
  console.log( propertyName in obj );                // true
  console.log( Object.hasOwn(obj, propertyName) );   // false
```
Which of these approaches suits your needs better?
For more info, you can look up `JS "in" operator vs Object.hasOwn`.

File: Sprint-2/implement/contains.test.js [hunk 2]
```diff
@@ -20,16 +20,27 @@ as the object doesn't contains a key of 'c'
...
-
+test("return false when the property does not exist", () =>{
+    expect(contains({a:1, b:3}, "c")).toEqual(false)
+});
 // Given invalid parameters like an array
 // When passed to contains
 // Then it should return false or throw an error
+test("return false when paramters invaliud ", () =>{
+    expect(contains([], "a")).toEqual(false)
+});
```
Comment (cjyuan) [inline Sprint-2/implement/contains.test.js]: This test does not yet confirm that the function correctly returns false when the first argument is an array. 
This is because `contains([], "a")` could also return `false` simply because "a" is not a key of the array.

Arrays are objects, with their indices acting as keys. A proper test should use a non-empty array along with a valid
key to ensure the function returns `false` specifically because the input is an array, not because the key is missing.

File: Sprint-2/implement/querystring.js [hunk 3]
```diff
@@ -1,13 +1,18 @@
...
     return queryParams;
   }
+
   const keyValuePairs = queryString.split("&");
 
   for (const pair of keyValuePairs) {
-    const [key, value] = pair.split("=");
-    queryParams[key] = value;
+    const [key, ...valueParts] = pair.split("=");
+    queryParams[key] = valueParts.join("=");
```
Comment (cjyuan) [inline Sprint-2/implement/querystring.js]: Note: (No change required)
- In real query string, both `key` and `value` are ***percent-encoded*** or ***URL encoded*** in the URL.
For example, the string "5%" is encoded as "5%25". So to get the actual value of "5%25"
(whether it is a key or value in the query string), we need to call a function to decode it.

- You can also explore the `URLSearchParams` API.

File: Sprint-2/implement/tally.js [hunk 4]
```diff
@@ -1,3 +1,26 @@
...
+  const result = {};
+
+  //  count items
+  for (const item of arr) {
+    if (result[item] === undefined) {
+      result[item] = 1;
+    } else {
+      result[item] = result[item] + 1;
+    }
+  }
```
Comment (cjyuan) [inline Sprint-2/implement/tally.js]: Does the following function call returns the value you expect?
```
tally(["toString", "toString"]);
```

Suggestion: Look up an approach to create an empty object with no inherited properties.

File: Sprint-2/interpret/invert.js [hunk 5]
```diff
@@ -5,25 +5,63 @@
...
 
   for (const [key, value] of Object.entries(obj)) {
-    invertedObj.key = value;
+    if (invertedObj[value] === undefined) {
+      invertedObj[value] = key;
+    } else if (Array.isArray(invertedObj[value])) {
+      invertedObj[value].push(key);
+    } else {
+      invertedObj[value] = [invertedObj[value], key];
+    }
```
Comment (cjyuan) [inline Sprint-2/interpret/invert.js]: Good job in recognising the possibility of collision and taking an extra step to address it.

File: Sprint-2/implement/contains.js (hunk 1 above)
Comment (MouawiaElkhalifa) [reply Sprint-2/implement/contains.js]: "Updated to use Object.hasOwn() and fixed the logic to correctly identify values inside arrays."

File: Sprint-2/implement/querystring.js (hunk 3 above)
Comment (MouawiaElkhalifa) [reply Sprint-2/implement/querystring.js]: Added the rest operator and .join("=") to prevent values with extra = signs from being cut off. Also used decodeURIComponent() to handle special characters like spaces."

File: Sprint-2/implement/tally.js (hunk 4 above)
Comment (MouawiaElkhalifa) [reply Sprint-2/implement/tally.js]: Switched to Object.create(null) to prevent prototype pollution. Added a test for reserved words like toString."

File: Sprint-2/interpret/invert.js (hunk 5 above)
Comment (MouawiaElkhalifa) [reply Sprint-2/interpret/invert.js]: Added collision handling so shared values are grouped into arrays. I also included Jest tests to verify this works for multiple duplicate values."

File: Sprint-2/interpret/invert.js [hunk 6]
```diff
@@ -1,29 +1,27 @@
...
-
-// b) What is the current return value when invert is called with { a: 1, b: 2 }
-
-// c) What is the target return value when invert is called with {a : 1, b: 2}
-
-// c) What does Object.entries return? Why is it needed in this program?
-
-// d) Explain why the current return value is different from the target output
-
-// e) Fix the implementation of invert (and write tests to prove it's fixed!)
```
Comment (cjyuan) [inline Sprint-2/interpret/invert.js:29]: You deleted your answers to these questions.

Anyway, do you recognise what this statement (in the original code) does?
>  invertedObj.key = value;

File: Sprint-2/implement/contains.test.js (hunk 2 above)
Comment (MouawiaElkhalifa) [reply Sprint-2/implement/contains.test.js]: Strengthened the array test to return false even if an index exists (e.g., contains(["a"], 0)).

File: Sprint-2/interpret/invert.js (hunk 6 above)
Comment (MouawiaElkhalifa) [reply Sprint-2/interpret/invert.js:29]: Restored missing answers (a-e) and corrected the logic. Replaced literal .key with dynamic [value] notation to prevent overwriting and ensure keys are swapped correctly.

File: Sprint-2/interpret/invert.js [hunk 7]
```diff
@@ -1,27 +1,42 @@
-/**
- * Invert Function
- * Swaps keys and values of an object. If multiple keys share the same value,
- * it groups those keys into an array to prevent overwriting data.
- */
+// a) What is the current return value when invert is called with { a : 1 }
+// {"1": "a"}
+
+// b) What is the current return value when invert is called with { a: 1, b: 2 }
+// {"1": "a", "2": "b"}
```
Comment (cjyuan) [inline Sprint-2/interpret/invert.js]: I think the "current return value" refers to the return value of the original code. 
The value you showed is the "target return value". 
I am sure you know the answers.

**Review by cjyuan — COMMENTED** [review]
Changes look good. Well done.

File: Sprint-2/interpret/invert.js (hunk 7 above)
Comment (MouawiaElkhalifa) [reply Sprint-2/interpret/invert.js]: thanks,Fixed (a) and (b) — they now show the correct current output.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
